File: 2020/go/day3.go
package advent2020

import (
	"fmt"
	"io/ioutil"
	"strings"
)

func Day3(inputFile string) error {
	input, err := ioutil.ReadFile(inputFile)
	if err != nil {
		return err
	}

	fmt.Println("Day 3: Toboggan Trajectory")
	part1, err := day3Part1(string(input))
	if err != nil {
		return err
	}
	fmt.Printf("\tPart 1: %d\n", part1)

	part2, err := day3Part2(string(input))
	if err != nil {
		return err
	}
	fmt.Printf("\tPart 2: %d\n", part2)

	return nil
}

func slopeTrees(rows []string, slope []int) int {
	x, trees := 0, 0
	r, d := slope[0], slope[1]
	for i := 0; i < len(rows); i += d {
		if x >= len(rows[i]) {
			x = x - len(rows[i])
		}
		if rows[i][x] == '#' {
			trees++
		}
		x += r
	}
	return trees
}

func day3Part1(input string) (int, error) {
	return slopeTrees(strings.Split(input, "\n"), []int{3, 1}), nil
}

func day3Part2(input string) (int, error) {
	acc := 1
	for _, s := range [][]int{{1, 1}, {3, 1}, {5, 1}, {7, 1}, {1, 2}} {
		acc *= slopeTrees(strings.Split(input, "\n"), s)
	}
	return acc, nil
}

File: 2020/rust/src/day3.rs
use std::error::Error;
use std::fs::File;
use std::io::prelude::*;

pub fn day3(input_file: &str) -> Result<(), Box<dyn Error>> {
    println!("Day 3: Toboggan Trajectory");

    let mut f = File::open(input_file)?;
    let mut input = String::new();
    f.read_to_string(&mut input)?;

    println!("\tPart 1: {}", part1(&input)?);
    println!("\tPart 2: {}", part2(&input)?);
    Ok(())
}

fn slope_trees(rows: Vec<Vec<char>>, slope: (usize, usize)) -> i32 {
    let (r, d) = slope;
    let (mut trees, mut x) = (0, 0);
    let mut rows = rows.into_iter().step_by(d);
    while let Some(row) = rows.next() {
        if x >= row.len() {
            x = x - row.len()
        }
        if row[x] == '#' {
            trees += 1;
        }
        x += r
    }
    trees
}

fn part1(input: &str) -> Result<i32, Box<dyn Error>> {
    Ok(slope_trees(
        input
            .lines()
            .map(|x| x.chars().collect())
            .collect::<Vec<Vec<char>>>(),
        (3, 1),
    ))
}

fn part2(input: &str) -> Result<i32, Box<dyn Error>> {
    let slopes: Vec<(usize, usize)> = vec![(1, 1), (3, 1), (5, 1), (7, 1), (1, 2)];
    Ok(slopes.iter().fold(1, |acc, cur| {
        acc * slope_trees(
            input
                .lines()
                .map(|x| x.chars().collect())
                .collect::<Vec<Vec<char>>>(),
            *cur,
        )
    }))
}
